Updated initial report for Cypherock X1 wallet (nov-22)

kartikCypherock opened the pull request.

kartikCypherock committed:
Updated initial report for Cypherock X1 wallet (nov-22)

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -35,33 +35,72 @@ social:
-In its [2020 Bitcointalk.org ANN post](https://bitcointalk.org/index.php?topic=5225800.0), Cypherock claims that the hardware wallet **will** be open source. It also makes this claim on [its website](https://twitter.com/BitcoinWalletz/status/1462700172719243266). The product is still in pre-order. We were able to locate Cypherock's Github page, but there was [no corresponding repository](https:// ...
+## Reproducibility of the release binary from the source code
+> - Cypherock X1 wallet firmware binary is signed by two private keys. The signature and a signed header is appended on top of firmware binary to form the complete Cypherock X1 wallet firmware image. 
+> - An important thing to note here is that the Cypherock X1 wallet firmware release binary is compiled using Docker, so, the binary will be reproducible only if the environment settings match. 
```
That's not clear. Docker is usually used precisely to improve the build environment's consistency. That phrase sounds like the opposite - like not using Docker would lead to more consistent builds.

kartikCypherock replied:
Hi Giszmo, the intention of the line was same as you mentioned. We use Docker to improve the consistency of builds. 
If anyone wants to reproduce the release binaries, they must create the same environment. 

I will reword this point so that it does not mislead anyone to think otherwise.

Giszmo reviewed (changes requested):
Overall the review is not finished. What's the verdict? Is it reproducible?

kartikCypherock committed:
Update MR based on review comments on line and well as offline. Updated the verdict:reproducible

Giszmo reviewed (changes requested):
Your verdict is not substantiated in the analysis. At some point it should show some form of:

Now with compiled.bin and downloaded.bin, we strip away the first 128B of signatures and hash them both:

```
$ sha256sum *.bin
b01221c3371aa9fb7a3135684c25b02b2945d434989b3bd7ce0305cfa75d5093  built_unsigned.bin
b01221c3371aa9fb7a3135684c25b02b2945d434989b3bd7ce0305cfa75d5093  compiled_unsigned.bin
```

which is what we want to see to give this product the verdict **reproducible**.

Also please mention the hash of the downloaded file for reference.

kartikCypherock committed:
Added compilation output using the steps for initial and main firmware binary. Also added fingerprint from released assets. Both signatures match.

kartikCypherock commented:
> Your verdict is not substantiated in the analysis. At some point it should show some form of:
> 
> Now with compiled.bin and downloaded.bin, we strip away the first 128B of signatures and hash them both:
> 
> ```
> $ sha256sum *.bin
> b01221c3371aa9fb7a3135684c25b02b2945d434989b3bd7ce0305cfa75d5093  built_unsigned.bin
> b01221c3371aa9fb7a3135684c25b02b2945d434989b3bd7ce0305cfa75d5093  compiled_unsigned.bin
> ```
> 
> which is what we want to see to give this product the verdict **reproducible**.
> 
> Also please mention the hash of the downloaded file for reference.

Hi @Giszmo, I have updated the report. Hopefully this should be sufficient. 
I have added few compilation steps, which leads to the binary. The hash matches with the downloaded file.

I have calculated the hash of the downloaded file as well which makes the comparison transparent.

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -94,6 +94,59 @@ Alternatively, one can follow below commands in a *powershell* window to build t
+        MINGW64 ~/Desktop/Cypherock/X1_wallet/Firmware/v0.4.773 ((v0.4.773))
+        $ sha256sum.exe main-release-outputs/release/Cypherock-Main.bin
+        54a2e548f3cb2f8a2f56671fe8b9df20ed90d084a9d68d435672e586e76c3903 *main-release-outputs/release/Cypherock-Main.bin
+***Signature of released unsigned main X1 Wallet firmware binary and compiled main X1 Wallet firmware binary matches.***
```
It's not the signature that needs to match.

Signature: The signer approves of the binary. Two signatures should never "match" as even the same signer would typically create different signatures given the same input.

Reproducible Build: The binary that may or may not be approved by one or more signers provably was compiled from the given source code.

Anyway, I can work with this now.

kartikCypherock replied:
I believe I misused the word signature here. 
The fingerprint (sha256 hash) of the binary matches.

kartikCypherock replied:
I have fixed it in another commit. Hope you don't mind.

kartikCypherock committed:
Reword signature with Fingerprint (sha256) to avoid confusion based on review comment

ujjwal-cyph commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,138 @@ providerWebsite: https://cypherock.com/
+### Reproducing the Cypherock X1 Wallet Initial firmware v0.4.773 on Windows 11
+* On Git Bash Shell, checkout using tag:
+        
+        MINGW64 ~/Desktop/Cypherock/X1_wallet/Firmware/v0.4.773 ((v0.4.773)) $ git clone git@github.com:Cypherock/x1_wallet_firmware.git .
+        MINGW64 ~/Desktop/Cypherock/X1_wallet/Firmware/v0.4.773 ((v0.4.773)) $ git checkout v0.4.773
```
make sure to pull submodule too.

```suggestion
        MINGW64 ~/Desktop/Cypherock/X1_wallet/Firmware/v0.4.773 ((v0.4.773)) $ git clone git@github.com:Cypherock/x1_wallet_firmware.git . --recurse-submodules
        MINGW64 ~/Desktop/Cypherock/X1_wallet/Firmware/v0.4.773 ((v0.4.773)) $ git checkout v0.4.773
```

Giszmo replied:
Is that why `ninja` failed? It should probably go into the documentation.

kartikCypherock replied:
Hi @Giszmo , unfortunately yes. I had the clone already on my environment so skipped this step.
I will update the document to reflect the same.

ujjwal-cyph commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,138 @@ providerWebsite: https://cypherock.com/
+### Reproducing the Cypherock X1 Wallet Main firmware v0.4.773 on Windows 11
+* On Git Bash Shell, checkout using tag:
+        
+        MINGW64 ~/Desktop/Cypherock/X1_wallet/Firmware/v0.4.773 ((v0.4.773)) $ git clone git@github.com:Cypherock/x1_wallet_firmware.git .
```
make sure to pull submodule too.

```suggestion
        MINGW64 ~/Desktop/Cypherock/X1_wallet/Firmware/v0.4.773 ((v0.4.773)) $ git clone git@github.com:Cypherock/x1_wallet_firmware.git . --recurse-submodules
```

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,138 @@ providerWebsite: https://cypherock.com/
+> - X1 Wallet supports upto 4 wallet recovery phrases, which can be used for transactions or even as a backup.
+> - X1 Wallet user can view the portfolio of all the 4 wallets from a single screen on their PC on the companion Cypherock CySync desktop application.
+> - X1 Wallet provides increased security against $5 wrench attack by distributing card into different geographical locations.
+> - X1 Wallet provides a 5 way navigation switch to perform user actions such as screen navigation and text input.
```
`>` denotes a quote and indeed, our analysis should quote some material by the provider. Please share where this list is from.

kartikCypherock replied:
These points are extrapolated from the website `https://www.cypherock.com/features` & `https://www.cypherock.com/security`

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,138 @@ providerWebsite: https://cypherock.com/
+> - X1 Wallet provides a 5 way navigation switch to perform user actions such as screen navigation and text input.
 
-> The Cypherock X1 uses 2 of 4 Shamir Secret Sharing (with the option to change the threshold in future) along with tamper-resistant hardware to secure the user’s private keys. Simply put, it is like Multi-sig but for the wallet recovery phrase.The seed phrase gets divided into 4 shards using the 2 of 4 SSS scheme and each of those shards gets stored in a unique cyCards. The cyCards can store upt ...
+From the [FAQ](https://www.cypherock.com/faq/): Can the company somehow steal the digital assets of the user?
+> - It is not possible for the company to steal the digital assets of the users. Cypherock X1 Wallet can recreate the private keys without internet access. The device needs to be connected to the internet just for broadcasting the transaction on the blockchain.
+> - Cypherock X1 Wallet is audited by [Keylabs](https://keylabs.io/), and the audit report can be found [here](https://dl.keylabs.io/keylabs_cypherock_x1_wallet_audit_final.pdf). Their report testifies that “The Cypherock X1 is an innovative wallet that uses many hardware and software security best practices and even features several security firsts that we have not yet seen in other wallets".
+> - As an organization, Cypherock  follows Kerckhoff's principle, and is therefore open source. Any user can download, audit the source code and design documents, and compile the firmware. 
+> - Cypherock X1 Wallet team is committed to providing an open source project in 2023 which can run on an NFC-enabled mobile device or on DIY hardware, by which a user can recover their seed phrase even if the company shuts down.
```
The quoted FAQ again is not backing up all the quotes. Only the first bullet point is the answer to the question in line 51.

I know you are a contributor to this product but WalletScrutiny is not a place to present original content.

kartikCypherock replied:
I understand, I have removed irrelevant points and stick to agenda.

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,138 @@ providerWebsite: https://cypherock.com/
+> - Cypherock X1 Wallet firmware is open sourced, hosted in GitHub repo: [Cypherock/x1_wallet_firmware](https://github.com/Cypherock/x1_wallet_firmware)
+> - Latest release source code: [Release v0.4.773](https://github.com/Cypherock/x1_wallet_firmware/releases/tag/v0.4.773)
+> - Cypherock X1 Wallet firmware is released in 2 flavors, “***initial firmware***” and “***main firmware***”.
+    > - ***Initial firmware*** is installed on the device during manufacturing. It provides minimum functionality such as X1 Wallet and X1 Card authentication which checks if the product is genuine or not. This firmware does not support any wallet related functionality. The initial firmware ceases to exist before any wallet activity is initiated.
+   > - ***Main firmware*** full fledged wallet related functionality and is installed on the device via the companion Cypherock CySync desktop application.
```
`>` denotes a quote in markdown, so this is all a quote. What's the source of the quote?

The varied indentation, with an extra `>` results in:

![Screenshot from 2022-12-15 20-47-13](https://user-images.githubusercontent.com/96378/207990455-6ec50019-8e43-410c-bdc3-9093f194e62a.png)

This should probably be:

```suggestion
    - ***Initial firmware*** is installed on the device during manufacturing. It provides minimum functionality such as X1 Wallet and X1 Card authentication which checks if the product is genuine or not. This firmware does not support any wallet related functionality. The initial firmware ceases to exist before any wallet activity is initiated.
    - ***Main firmware*** full fledged wallet related functionality and is installed on the device via the companion Cypherock CySync desktop application.
```

kartikCypherock replied:
Yes, apologies for these formatting errors, I have removed all `>` as there is no quotations required.

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,138 @@ providerWebsite: https://cypherock.com/
 
-## Private keys are not shared 
+## Reproducibility of the release binary from the source code
+> - Cypherock X1 Wallet firmware binary is signed by two private keys. The signature and a signed header is appended on top of firmware binary to form the complete Cypherock X1 Wallet firmware image. 
+> - An important thing to note here is that the Cypherock X1 Wallet firmware release binary is compiled using Docker. 
+> - Fingerprint (SHA256 digest) of the latest Cypherock X1 Wallet firmware binary (unsigned) can be found on the release page: [Release v0.4.773](https://github.com/Cypherock/x1_wallet_firmware/releases/tag/v0.4.773)
+> - The docker file corresponding to the latest release can be found here: [Dockerfile v0.4.773](https://github.com/Cypherock/x1_wallet_firmware/blob/v0.4.773/Dockerfile)
+> - List of tools required to successfully build the project can be found here: [Building project](https://github.com/Cypherock/x1_wallet_firmware/tree/v0.4.773#building-project)
+> - Sequence of commands run on docker can be found here: [Docker commands v0.4.773](https://github.com/Cypherock/x1_wallet_firmware/blob/v0.4.773/.github/workflows/containerized-build.yml)
```
Again: What are you quoting here?

kartikCypherock replied:
Apologies for these formatting errors, I have removed all `>` as there is no quotations required.

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,138 @@ providerWebsite: https://cypherock.com/
+
+* On Powershell, setup Docker and compile:
+
+        PS C:\Users\LENEVO\Desktop\Cypherock\X1_wallet\Firmware\v0.4.773>  docker run -v ${pwd}/build:/out -it x1-wallet-app-env /bin/ash docker build . --file Dockerfile --tag x1-wallet-app-env
```
This is two docker commands in one line, probably in reverse order.

kartikCypherock replied:
I have fixed the order.

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,138 @@ providerWebsite: https://cypherock.com/
+        .
+        .
+        /home/build/Release # cmake -DDEV_SWITCH=OFF -DDEBUG_SWITCH=OFF -DSIGN_BINARY=OFF -DCMAKE_BUILD_TYPE:STRING=Release -DFIRMWARE_TYPE=Initial -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=ON -DCMAKE_BUILD_PLATFORM:STRING=Device -G Ninja ../../
+        /home/build/Release # ninja
```
`ninja` fails for me. I had tried it first with the first  code block and now again with this one with the submodules but both times I get:

```
...
   72 | #include "lv_port_disp.h"
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
ninja: job failed: /usr/bin/arm-none-eabi-gcc -DFIRMWARE_HASH_CALC=1 -DSTM32L4 -DSTM32L486xx -DUSBD_SOF_DISABLED -DUSE_BIP32_CACHE=0 -DUSE_BIP39_CACHE=0 -DUSE_HAL_DRIVER -DUSE_SIMULATOR=0 -DX1WALLET_INITIAL=1 -DX1WALLET_MAIN=0 -I../../initial/config -I../../src -I../../src/level_one/controller -I../../src/level_one/tasks -I../../src/level_two/controller -I../../src/level_two/tasks -I../../src/level_three/add_wallet/controller -I../../src/level_three/add_wallet/tasks -I../../src/level_three/advanced_settings/controller -I../../src/level_three/advanced_settings/tasks -I../../src/level_three/old_wallet/controller -I../../src/level_three/old_wallet/tasks -I../../src/level_four/core/controller -I../../src/level_four/core/tasks -I../../src/level_four/card_health_check -I../../src/level_four/factory_reset -I../../src/level_four/tap_cards/controller -I../../src/level_four/tap_cards/tasks -I../../common/interfaces/card_interface -I../../common/interfaces/desktop_app_interface -I../../common/interfaces/flash_interface -I../../common/interfaces/user_interface -I../../common/libraries/atecc -I../../common/libraries/atecc/atcacert -I../../common/libraries/atecc/basic -I../../common/libraries/atecc/crypto -I../../common/libraries/atecc/crypto/hashes -I../../common/libraries/atecc/hal -I../../common/libraries/atecc/host -I../../common/libraries/atecc/jwt -I../../common/libraries/crypto -I../../common/libraries/crypto/aes -I../../common/libraries/crypto/chacha20poly1305 -I../../common/libraries/crypto/ed25519-donna -I../../common/libraries/crypto/monero -I../../common/libraries/crypto/random_gen -I../../common/libraries/proof_of_work -I../../common/libraries/shamir -I../../common/libraries/util -I../../common/startup -I../../common/logger -I../../common/coin_support -I../../common/flash -I../../common/Firewall -I../../common -I../../common/lvgl -I../../common/lvgl/porting -I../../common/lvgl/src -I../../common/lvgl/src/lv_core -I../../common/lvgl/src/lv_draw -I../../common/lvgl/src/lv_font -I../../common/lvgl/src/lv_hal -I../../common/lvgl/src/lv_misc -I../../common/lvgl/src/lv_objx -I../../common/lvgl/src/lv_themes -I../../stm32-hal -I../../stm32-hal/BSP -I../../stm32-hal/Inc -I../../stm32-hal/Drivers/CMSIS/Include -I../../stm32-hal/Drivers/CMSIS/Device/ST/STM32L4xx/Include -I../../stm32-hal/Drivers/STM32L4xx_HAL_Driver/Inc -I../../stm32-hal/Drivers/STM32L4xx_HAL_Driver/Inc/Legacy -I../../stm32-hal/Drivers/STM32L4xx_HAL_Driver/Src -I../../stm32-hal/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Inc -I../../stm32-hal/Middlewares/ST/STM32_USB_Device_Library/Class/CDC/Src -I../../stm32-hal/Middlewares/ST/STM32_USB_Device_Library/Core/Inc -I../../stm32-hal/Middlewares/ST/STM32_USB_Device_Library/Core/Src -I../../stm32-hal/Peripherals -I../../stm32-hal/Peripherals/Buzzer -I../../stm32-hal/Peripherals/display -I../../stm32-hal/Peripherals/display/SSD1306 -I../../stm32-hal/Peripherals/flash -I../../stm32-hal/Peripherals/logger -I../../stm32-hal/Peripherals/nfc -I../../stm32-hal/Src -I../../stm32-hal/Startup -I../../stm32-hal/porting -I../../stm32-hal/libusb -I../../stm32-hal/libusb/inc -I../../stm32-hal/libusb/src -O3 -DNDEBUG -mcpu=cortex-m4 -mthumb -mfpu=fpv4-sp-d16 -mfloat-abi=hard -fdata-sections -ffunction-sections -Wall -Wno-format-truncation -Wno-unused-but-set-variable -Wno-return-type -D_POSIX_C_SOURCE=200809L -std=c11 -MD -MT CMakeFiles/Cypherock-Initial.elf.dir/src/level_one/controller/controller_level_one_initial.c.obj -MF CMakeFiles/Cypherock-Initial.elf.dir/src/level_one/controller/controller_level_one_initial.c.obj.d -o CMakeFiles/Cypherock-Initial.elf.dir/src/level_one/controller/controller_level_one_initial.c.obj -c ../../src/level_one/controller/controller_level_one_initial.c
In file included from ../../common/interfaces/user_interface/ui_common.h:13,
                 from ../../common/interfaces/user_interface/ui_menu.h:13,
                 from ../../src/constant_texts.h:13,
                 from ../../src/controller_main.h:32,
                 from ../../src/level_one/controller/controller_level_one.h:14,
                 from ../../src/level_one/controller/controller_level_one_initial.c:60:
../../common/lvgl/lvgl.h:72:10: fatal error: lv_port_disp.h: No such file or directory
   72 | #include "lv_port_disp.h"
      |          ^~~~~~~~~~~~~~~~
compilation terminated.
ninja: subcommands failed
```

I do not have a windows system but I doubt that's the issue.

kartikCypherock replied:
It failed due to incomplete clone of the git repository. You need to use `--recurse-submodules` switch with the git command. 
I have updated the commands.

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,138 @@ providerWebsite: https://cypherock.com/
-But with that said, this hardware wallet has not yet been released.
\ No newline at end of file
+## Installing firmware updates
+> - The user has the flexibility to install a firmware update or not. The secure bootloader, which forms the root of trust, verifies if the firmware is a newer version than the currently installed version and [verifies the source of the firmware](https://github.com/Cypherock/x1_wallet_firmware/blob/main/docs/bootloader.md) (through 2 signatures).
```
again a quote without source.

kartikCypherock replied:
Quotation mark not required. Source is already included with a hyperlink

kartikCypherock committed:
Fix document to fix formatting issues and compilation error due to incomplete checkout

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -6,8 +6,8 @@ authors:
 - danny
```
Please add yourself to the authors here. You may either add yourself with a full name:

```
- John Doe
```

which would get added to the authors list without link or you add an authors file to /_authors like frequent contributors did.

Also please, if you are a contributor to the product, add a disclaimer.

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,135 @@ providerWebsite: https://cypherock.com/
+- X1 Wallet provides a 5 way navigation switch to perform user actions such as screen navigation and text input.
 
-> The Cypherock X1 uses 2 of 4 Shamir Secret Sharing (with the option to change the threshold in future) along with tamper-resistant hardware to secure the user’s private keys. Simply put, it is like Multi-sig but for the wallet recovery phrase.The seed phrase gets divided into 4 shards using the 2 of 4 SSS scheme and each of those shards gets stored in a unique cyCards. The cyCards can store upt ...
+From the [FAQ](https://www.cypherock.com/faq/): Can the company somehow steal the digital assets of the user?
+- It is not possible for the company to steal the digital assets of the users. Cypherock X1 Wallet can recreate the private keys without internet access. The device needs to be connected to the internet just for broadcasting the transaction on the blockchain.
```
Mind adding something like this?

> The provider [further claims](https://github.com/Cypherock/x1_wallet_firmware/blob/main/docs/shamir_secret_sharing.md#compatibility-with-bip39) to use entropy from the ATECC608A "secure element" which is not open to public audit but also from the microprocessor which runs the open source code and thus should be able to prevent poor entropy backdoors in the closed source of the ATECC608A

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,135 @@ providerWebsite: https://cypherock.com/
 
-## Private keys are not shared 
+## Reproducibility of the release binary from the source code
+- Cypherock X1 Wallet firmware binary is signed by two private keys. The signature and a signed header is appended on top of firmware binary to form the complete Cypherock X1 Wallet firmware image. 
```
I assume the HWW checks these signatures in its bootloader whenever it's started?

Edit: further down this is claimed, right?

kartikCypherock replied:
Yes, HWW verifies the signature of the header and the firmware, only then the firmware installation process is considered to be complete.

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -6,8 +6,8 @@ authors:
-version: 
+updated: 2022-12-5
+version: 0.4.773
 binaries: 
```
```suggestion
binaries: https://github.com/Cypherock/x1_wallet_firmware/releases
```

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -6,8 +6,8 @@ authors:
 - danny
 released: 
```
A date would be nice to have here. When was the first unit sold? Or the date of the first production firmware?

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -6,8 +6,8 @@ authors:
 discontinued: 
-updated: 
-version: 
+updated: 2022-12-5
```
This "updated" refers to when the software was updated. On the [release page](https://github.com/Cypherock/x1_wallet_firmware/releases/tag/v0.4.773) I see:

```suggestion
updated: 2022-11-29
```

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,135 @@ providerWebsite: https://cypherock.com/
+### Alternatively, one can follow below commands in a *powershell* window to build the Cypherock X1 Wallet firmware:
+* For Initial firmware - unsigned binary will be generated at build/initial/Cypherock-Initial.bin
+        
+        docker build . --file Dockerfile --tag x1-wallet-app-env
+        mkdir -p build
+        docker run -v ${pwd}/build:/out -it x1-wallet-app-env /bin/ash
```
This documentation lacks context. Up to here there was no mention on what to clone.
These first three lines are the same as for the "Main" firmware and should be prepended by these three lines:

```
        git clone https://github.com/Cypherock/x1_wallet_firmware
        cd x1_wallet_firmware
        git submodule update --init --recursive
        docker build . --file Dockerfile --tag x1-wallet-app-env
        mkdir -p build
        docker run -v ${pwd}/build:/out -it x1-wallet-app-env /bin/ash
```

kartikCypherock replied:
Both Initial and Main firmware are maintained in the same repository and are released together.

I have re-structured the document to minimize duplication of text and solve this issue.

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,135 @@ providerWebsite: https://cypherock.com/
+### Alternatively, one can follow below commands in a *powershell* window to build the Cypherock X1 Wallet firmware:
+* For Initial firmware - unsigned binary will be generated at build/initial/Cypherock-Initial.bin
+        
+        docker build . --file Dockerfile --tag x1-wallet-app-env
+        mkdir -p build
+        docker run -v ${pwd}/build:/out -it x1-wallet-app-env /bin/ash
```
docker should be run with the `--rm` flag to not clutter the user's disk. The commands can easily be chained. How about:

```
docker run --volume=${pwd}/build:/out --rm -it x1-wallet-app-env /bin/ash -c "\
    mkdir -p /dist/initial && cd /home/build/Release/ ;\
    cmake -DDEV_SWITCH=OFF -DDEBUG_SWITCH=OFF -DSIGN_BINARY=OFF -DCMAKE_BUILD_TYPE:STRING=Release -DFIRMWARE_TYPE=Initial -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=ON -DCMAKE_BUILD_PLATFORM:STRING=Device -G Ninja ../../ ; \
    ninja; \
    cp Cypherock-*.* /dist/initial && cp -a /dist/. /out;\
    /bin/ash"
exit
```

I started `ash` in the end so the container remains open for inspection should anything go wrong. There's probably more elegant ways. For compile instructions, assuming all goes well, this would be shorter:

```
docker run --volume=${pwd}/build:/out --rm -it x1-wallet-app-env /bin/ash -c "\
    mkdir -p /dist/initial && cd /home/build/Release/ ;\
    cmake -DDEV_SWITCH=OFF -DDEBUG_SWITCH=OFF -DSIGN_BINARY=OFF -DCMAKE_BUILD_TYPE:STRING=Release -DFIRMWARE_TYPE=Initial -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=ON -DCMAKE_BUILD_PLATFORM:STRING=Device -G Ninja ../../ ; \
    ninja; \
    cp Cypherock-*.* /dist/initial && cp -a /dist/. /out"
```

kartikCypherock replied:
We will improve this in our build system as a separate task as this requires some experiments. Thank you for the initial pointers!

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,135 @@ providerWebsite: https://cypherock.com/
+        .
+
+* On Powershell, setup Docker and compile:
+        PS C:\Users\LENEVO\Desktop\Cypherock\X1_wallet\Firmware\v0.4.773>  docker build . --file Dockerfile --tag x1-wallet-app-env
```
Omitting an empty line results in:
![Screenshot from 2023-01-04 22-16-50](https://user-images.githubusercontent.com/96378/210679878-aeff865b-ffb7-4b1c-a457-c8c0e4763ec9.png)


```suggestion
* On Powershell, setup Docker and compile:

        PS C:\Users\LENEVO\Desktop\Cypherock\X1_wallet\Firmware\v0.4.773>  docker build . --file Dockerfile --tag x1-wallet-app-env
```

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,135 @@ providerWebsite: https://cypherock.com/
 
-## Interface
+* On Powershell, setup Docker and compile:
+        PS C:\Users\LENEVO\Desktop\Cypherock\X1_wallet\Firmware\v0.4.773>  docker build . --file Dockerfile --tag x1-wallet-app-env
```
```suggestion
* On Powershell, setup Docker and compile:

        PS C:\Users\LENEVO\Desktop\Cypherock\X1_wallet\Firmware\v0.4.773>  docker build . --file Dockerfile --tag x1-wallet-app-env
```

Giszmo commented on _hardware/cypherockx1.md:
```diff
@@ -16,52 +16,135 @@ providerWebsite: https://cypherock.com/
\ No newline at end of file
+## Installing firmware updates
+- The user has the flexibility to install a firmware update or not. The secure bootloader, which forms the root of trust, verifies if the firmware is a newer version than the currently installed version and [verifies the source of the firmware](https://github.com/Cypherock/x1_wallet_firmware/blob/main/docs/bootloader.md) (through 2 signatures).
+- Users can view the installed version and the commit hash of the release corresponding to the installed Cypherock X1 Wallet firmware binary and can verify it against the GitHub repository. This cross check should provide enough confidence to the user that the installed firmware binary matches the release binary on the public source. 
```
A git hash is just a claim as it can't be verified from the binary. The below mentioned display of the binary's hash is essential to prevent binaries claiming to use a certain code revision to be deployed.

We currently handle this inconsistently but plan to be strict about hardware wallets displaying binary hashes.

kartikCypherock replied:
I understand that the git hash does not help the X1 wallet claim that the deployed firmware is actually running or not.

As mentioned in the 3rd point, we are planning to add functionality in the X1 wallet bootloader through which users can see the firmware hash after a successful firmware update.

Giszmo reviewed (changes requested):
I didn't get to actually compare hashes. For some reason the artefact doesn't come out of the docker container. Just dropping the other issues here meanwhile ...

kartikCypherock commented:
> I didn't get to actually compare hashes. For some reason the artefact doesn't come out of the docker container. Just dropping the other issues here meanwhile ...

Hi @Giszmo,

Above set of commands work for me. I am using Powershell on Windows 11 machine.

The last command `cp Cypherock-*.* /dist/main && cp -a /dist/. /out` brings the artefacts out of the docker container.
Can you please ensure that you ran this command after compiling the source code?

Please let me know your setup in case it still doesn't work for you.

kartikCypherock committed:
Add details regarding release dates, update document to reflect latest reproducibility of latest production release (Jan-23), fix review comments

Giszmo commented:
My apologies. The powershell appears to work slightly different than the linux shell while simultaneously not complaining about the "wrong" use. `${pwd}` resolved to `""` and the folder was created in the root `/build` instead of complaining about its absence. Now things worked better.

kartikCypherock commented:
> My apologies. The powershell appears to work slightly different than the linux shell while simultaneously not complaining about the "wrong" use. `${pwd}` resolved to `""` and the folder was created in the root `/build` instead of complaining about its absence. Now things worked better.

Hi, ok good to know that you were able to replicate the steps and get the binaries.
Did you try for both X1 wallet initial and main firmware?
